"What is Jupyter" explainer document (#728)
jupyter/jupyter: docs/source/what_is_jupyter.md

program outputs, and metadata inside a structured text file.

### JupyterHub

- [JupyterHub for multi-user setups](https://jupyterhub.readthedocs.io)

JupyterHub provides a multi-user management system where many different people
can log in and use their own isolated notebook editor program and environment.

With JupyterHub, editing and running notebook files is still performed by an
editor program like Lab or Notebook. The Hub is responsible for authenticating
users and providing them their corresponding Lab/Notebook instance connection.

JupyterHub can be set up in different system configurations:

- In a single computer
- In a cluster of computers
- In containers in the cloud

JupyterHub is quite flexible for many different multi-user scenarios.

### Interactive Buttons, Sliders and more with ipywidgets

- [Jupyter Widgets (AKA ipywidgets)](https://ipywidgets.readthedocs.io/en/latest/)

Jupyter Widgets (also called ipywidgets) provides interactive interface
elements that you can add directly into your notebooks.

- Note: The subproject and high-level references to the widgets were renamed to
Jupyter-Widgets, but Jupyter-Widgets and IPyWidgets are the very same thing. To

> not break compatibility with older code, the software library is still called

Author: Updated! Thanks for the suggestion here.

# What is Jupyter?

"Jupyter" can mean a lot of things, so let's start from the beginning, and

> break it down:

Author: Sounds good, I'll refactor this with more inclusive language.

> break it down:

Author: Done, see a83604f2bd10

> First and foremost, "Project Jupyter" is a large umbrella project that covers

Author: I'll work on a refactor for this.

> First and foremost, "Project Jupyter" is a large umbrella project that covers

Author: Done, see a83604f2bd10

many different software offerings and tools. That includes "Jupyter Notebook"

> and "JupyterLab", which are both popular notebook-editor programs. **The**

Author: See goals from [this comment.](https://github.com/jupyter/jupyter/pull/728#discussion_r1613573049)

**Jupyter project, and its subprojects, all center around providing tools**
**(and standards) for interactive computing with computational notebooks.**
If you're not familiar with some of those things, you can take a guided tour
below that explains each one from the start.

The term "Jupyter" is often used as a shorthand to refer to one of those
products or ideas, which can cause confusion. Let's take a look at each piece,
and provide some clarity, below.

> # What is a "Computational Notebook" anyway?

Reviewer: ```suggestion
# What is a computational notebook?
```

"Computational Notebook" is not a proper noun. I rarely see the term "computational" used — should we just use "notebook" (in lowercase) here?

Author: Resolved in another comment.

> A famous computer programmer (Donald Knuth) popularized the idea to combine
> explanatory plain english text with computer code, which is commonly called

Reviewer: ```suggestion
Donald Knuth popularized the idea of combining 
explanatory plain English text with computer code, which is commonly called
```

Mr. Knuth needs no introduction 😄

> "literate programming". **By adopting this practice, computer programs, as**
> **well as other complex information and ideas, could be better explained to**
> **a wide range of people.**

Reviewer: ```suggestion
"literate programming". Computer programs, as
well as other complex information and ideas, can be explained to
a wide range of people using literate programming.
```

Bold not necessary here

A program written in this way could be printed on paper and shared by hand
as an actual "notebook", but in modern times, they are shared digitally as
"notebook files", and can contain additional rich media like images, 3D models
and interactive figures, along with data and other program outputs.

By opening a notebook with an editor program like JupyterLab, you can also
run the code inside the notebook. Since a notebook can contain code that does
virtually anything, you can do nearly anything regular software can do inside
of a notebook. For example, a notebook file might be used to:

- Read spreadsheets (or create them) to build reports about your household spending
- Show proof of your college thesis with interactive graphs and source data
- Process astronomical data from the James Webb Space Telescope (JWST) (CHECK/LINK)
- Generate an image of a black hole by processing telescope data (CHECK/LINK)
- Calculate the presence of gravitational waves from observatory data (CHECK/LINK)

Those last three are real world examples that demonstrate the scientific

> community's usage of computational notebooks (CHECK). Because scientists,

Reviewer: Probably not necessary to say that these are real-world examples

engineers, and other technical people so often need to perform and communicate

> sophisticated calculations to describe their work, these "computational
> notebooks" became a very popular way to share their work and ideas.

Reviewer: ```suggestion
sophisticated calculations to describe their work, these computational
notebooks became a very popular way to share their work and ideas.
```

Not necessary to quote the term; boldface might be useful to highlight it on first use, if you do so consistently

> Notebooks are not just for Nasa scientists either: Students, hobbyists, and

Reviewer: Capitalize NASA or omit it

business people commonly use the fast, interactive workflows in JupyterLab to
make notebooks that help solve everyday problems. You can use them to explain
your own ideas to others, to learn, to automate tasks at home or at work,

> visualize complex information, and more.

Reviewer: Either use "to" to begin each clause, including "visualize …", or delete it from all clauses (and perhaps delete "learn" for being too terse and generic). I historically have favored using "to" a lot, although an editor I worked with talked me out of using it in more conversational, casual writing styles.

When someone uses the term "notebook", they might be referring to:

- A notebook file on their computer
- The idea of combining computer code, explanatory text, images and more into
  the "notebook format"
- The "Jupyter Notebook" application, used to author and edit digital notebook
  files
- Jupyter's ``.ipynb`` notebook file format (used to save your notebook files

>   on your computer), which is interpreted by the ``nbformat`` software library

Reviewer: 1 and 3 are the most common uses of "notebook" that new users need to be aware of. I don't think newbies are going to be as concerned with discussing the format itself. OK to delete points 2 and 4 from an introductory doc.

And the term "Jupyter" might refer to:

- "Project Jupyter", the overarching umbrella project
- The "Jupyter Notebook" or "JupyterLab" editing programs (or other Jupyter
  products)

The name Jupyter comes from the three programming languages the project
originally supported: Julia (ju), Python (pyt) and R (r).

> # How do the Jupyter notebook-editor programs work?
> 
> It is common for notebook-editor programs like JupyterLab, or Jupyter

Reviewer: ```suggestion
# How do Jupyter notebook editors work?

It is common for notebook editors like JupyterLab, or Jupyter
```

I think "editor program" is overly verbose; most people with text or code writing experience will understand "editor"

Notebook, to share some features and workflows, because they are influenced
by a common set of ideas about computational notebooks, their advantages,
and how best to work with them effectively.

> Let's break down some of those ideas.

Reviewer: Recommend something other than "break down", which is idiomatic and ambiguous

## Interactive programming (the REPL)

In the past, writing programs, running them, and seeing results was commonly
a slower and more deliberative process than it is today.

As time passed, new techniques were discovered that sped up the process between
writing code, running it, and seeing results. Languages like Python were also
later introduced that offered some unique advantages over older languages.

For instance, Python is an interpreted language, so it does not need to be
compiled (LINK) before it can be run (saving programmers steps and time).
It is also a dynamically-typed language that does not require the programmer
to specify the type of their data ahead of time before using it, which can
sometimes save time and reduce the complexity of code (particularly for
smaller, simpler programs).

> 

Reviewer: I think that this introduction is less about Jupyter and more about Python. Probably OK to delete, to get right to the point about the REPL, and specifically to say what Jupyter does with the REPL.

Another programming technique (which is key to the Jupyter notebook-editing
programs) is the Read-Eval-Print-Loop (REPL), which allows a programmer to
interactively write code, run it, keep their data and variables intact, then
rewrite, re-run, and refine their code on-the-fly (without losing data
after the program finishes running).

A REPL is so named because the programmer writes snippets of code that are
first read (R), then Evaluated (E) or in other words executed, the results
are printed (P) to some kind of display or output, and that process happens
repeatedly in a loop (L), where the REPL waits until the programmer has
another snippet of code to execute.

New snippets of code can refer to variables defined in previous Eval steps,
because the REPL keeps objects that were created (by previous runs of the

> loop) in-memory, until the user closes their REPL.

Reviewer: This should be rewritten to talk about Jupyter's use of the REPL. If there isn't enough info that's Jupyter-specific, we could delete this section for brevity.

## Kernels

> Notebook editor programs like JupyterLab create a REPL (read more about those
> above) for each of your open notebook files, in a language of your choice. In
> the Jupyter ecosystem, each of these REPL's is called a kernel, and it holds
> the data and objects you create with your notebook code in a long-running
> program on your computer (the management and creation of these kernels
> is orchestrated by the Jupyter Server (LINK), more on that later).

Reviewer: ```suggestion
Notebook editors, like JupyterLab, create a REPL for each of your open notebook files, in the language of your choice. In
the Jupyter ecosystem, each of these REPLs is called a **kernel**, and it holds
the data and objects you create with your notebook code in a long-running
program on your computer.
```

If the above section is deleted, consider linking REPL to some third-party definition, such as on python dot org.

We don't need to bring up Jupyter Server at this stage of an introduction, especially not in a parenthetical.

> This is why you don't lose your data and variables when you execute multiple
> notebook cells in a row, they stay alive inside the REPL so that you can
> continue to use them to explore the data and problem you are investigating
> (though if you shut down the Jupyter server that holds the kernels, you will
> lose those variables).

Reviewer: ```suggestion
The REPL keeps your data and variables in memory when you execute multiple
notebook cells in a row. If you shut down or restart your kernel, you will
lose those variables,
and you'll need to run the code in your notebook again to get them back.
```

Again, too early to talk about servers; let's keep the kernel as the key concept in this section

The kernels that run in the background for each of your notebooks are what
power the fast, exploratory programming workflows that Jupyter notebook

> editor programs excel at.

Reviewer: Not sure if this is necessary in a technical introduction; this reads less like relevant information and more like promotional material

Because Jupyter Server provides kernel communication and management features
in a cohesive, self-contained package, for instance, new notebook editor
programs can focus solely on adding new interface and editing experiences,
leaving the task of creating and managing new kernels completely up to the
Jupyter Server.

JupyterCAD is a 3D modeling tool built on top of Jupyter software, for
instance, and there are other examples too (LINK HERE).

## Talking to Kernels (The Jupyter Protocol)

Anyone can make new software that talks to Jupyter's kernels (read more
about those above), by using the Jupyter Protocol. The protocol provides
a standardized blueprint for passing information back and forth between
the kernels and other software.

By implementing the designs described in the Jupyter Protocol, you could
invent a completely new interactive programming experience, or add support
for a new programming language to Jupyter.

Because Project Jupyter is free and open, it encourages anyone to explore
new ways of working with their notebooks and kernels, and likes to offer
compatibility and interoperation with other software.

TODO add more here? examples? etc.

> 

Reviewer: I'm concerned that this section is getting into significant technical detail that is not necessary for people just getting started with Jupyter. I feel like this document should be "100-level" (in US university terms), but this block is 200-level information, for people who want to know more.

# What else should I know about Project Jupyter?

Jupyter software is free and open-source, developed by a global community
of volunteers and contributors, available for the benefit of all.

Project Jupyter welcomes people from all backgrounds and with many types
of skills (not just software!) so we encourage you to join us!

You can share feedback about your experiences directly with the people who
make Jupyter software, or volunteer and contribute to help Jupyter in many
different ways, like:

- Writing tutorials
- Testing newly released versions
- Adding software features
- Hosting the weekly video meetings (LINK)
- Helping others in the community
- ...and more!

The project is split into largely independent subprojects which handle
different aspects of Jupyter software and the community. A central council,
the Jupyter Executive Council (LINK), makes decisions about project-wide
goals and policies, while different subprojects handle the actual development

> of the various software components.

Reviewer: This may be too much detail for an introduction. Might be best to link to the governance docs, or perhaps to delete this entirely.

            click Nbformat href "https://nbformat.readthedocs.io" _blank

        JupyterWidgets>Jupyter-Widgets]
            click JupyterWidgets href "http://ipywidgets.readthedocs.io" _blank
    end

    subgraph frontends
        direction LR

        JupyterLab(Jupyter-Lab)
            click JupyterLab href "https://jupyterlab.readthedocs.io" _blank

        JupyterNotebook(Jupyter-Notebook)
            click JupyterNotebook href "https://jupyter-notebook.readthedocs.io" _blank
    end

    subgraph viewers
        direction LR

        Binder(Binder)
            click Binder href "https://mybinder.org" _blank

        Voila(Voilà)
            click Voila href "https://voila.readthedocs.io" _blank

        Nbviewer[nbviewer]
            click Nbviewer href "https://nbviewer.org" _blank

        Nbconvert[nbconvert]
            click Nbconvert href "https://nbconvert.readthedocs.io" _blank
    end

    JupyterHub(Jupyter-Hub)
        click JupyterHub href "https://jupyterhub.readthedocs.io" _blank

    JupyterHub --> frontends
    IPython ~~~ source/code
    frontends --> source/code
    viewers --> source/code
```

### Notebook Editor Programs

> - [JupyterLab](https://jupyterlab.readthedocs.io/en/latest/)

Reviewer: Is it too much to characterize JupyterLab as an IDE in this section?

Author: Going by the target audience again (maybe a chemist who does a little python for their lab work), I think this is unnecessary jargon that would hurt communication: it introduces a new idea, a complex one, and throws uncertainty onto the nature of JupyterLab, where I think editor is a plain english alternative that works better here.

- [Jupyter Notebook](https://jupyter-notebook.readthedocs.io/en/stable/)

The two most popular notebook editor programs are Jupyter-Notebook and

> JupyterLab, though there are others. Which one you choose to use is

Reviewer: Rather than describing these as "the two most popular notebook editors", we might want to instead characterize them as two notebook editors **published by Project Jupyter**. There are many other editors produced and published by 3rd-party commercial vendors, including my own employer, and which may be popular. We can't endorse any of them in this document, but I am concerned that this phrasing seems to minimize the outside developer community for notebook editors.

Author: I agree the third parts editor are important and would like to emphasize that here. I'll refactor this and add those in.

Author: Resolved in 300b0ffd4

mostly a matter of preference (both Notebook and Lab provide many of the
same capabilities for editing and running notebooks):

- Jupyter Notebook offers a simplified, lightweight notebook authoring
  experience

- JupyterLab offers a more customizable, feature-rich, tabbed multi-notebook
  editing environment, and has additional tools like a customizable layout
  and system console (a common tool used by programmers)

- And more...read about additional notebook interfaces here (LINK)!

A professor, for instance, might use JupyterLab for their daily work as a data
analyst, and Jupyter Notebook while teaching to provide a simpler, more focused
view of the notebook's content to their students.

### The notebook format

- [nbformat.readthedocs.io](https://nbformat.readthedocs.io)

Jupyter uses the .ipynb format to store your notebook files (where the data
and outputs of your notebook reside for long term storage). The .ipynb format
is interpreted and modified by the nbformat software library.

> TODO: features, more info, basic exmaplanation of how it works/structure

Reviewer: Not sure if this is necessary in a basic introduction.

Author: So to elaborate on some principles guiding this document:

The primary goal is to clarify and disambiguate essential ideas and components of the Jupyter ecosystem.

The notebook editor programs are used as an organizing pillar in this document because conceptually they're easy to apprehend (given their similarities to common document editor programs) and can serve as a central conceptual hub to build additional concepts and understanding onto.

The notebook format needs to be included because it contributes to confusion around the term "notebook", and because it is a foundational piece underlying Jupyter/the editors. It is included because users should be clear on the difference between Notebook (the Editor), notebook (a notebook file), computational notebooks (the notebook concept based on literate programming) and notebook (the .ipynb file format).

Eliminating confusion around those concepts is a key goal, because these issues form a big part of users' uncertainty about which pieces are there and what they do.

To be clear, this document should be something users can read early on in their Jupyter experience. I have confidence that, with plain language, and clear guidance, our users are capable of digesting each of these pieces.

A long time ago, a precursor to Jupyter was created, called ipython (see
the [History of Jupyter][] to learn more about that). It provided some of
the same fast, advanced REPL features that Jupyter still has today, and it
is the default Python kernel bundled with JupyterLab and Notebook. Some of
the features ipython provides:

- Foobar
- Bizbaz
- WikRakRum
- interactive shell: [ipython.readthedocs.io/interactive/tutorial][]
- magic commands: [ipython.readthedocs.io/interactive/magics][]

For more details on IPython features check [`docs.jupyter.org > Projects > IPython`](https://docs.jupyter.org/en/latest/projects/ipython_projects.html).

### Interactive Buttons, Sliders and more with ipywidgets

- [Jupyter Widgets (AKA ipywidgets)](https://ipywidgets.readthedocs.io/en/latest/)

Jupyter Widgets (also called ipywidgets) provides interactive interface
elements that you can add directly into your notebooks.

- Note: The subproject and high-level references to the widgets were renamed to
Jupyter-Widgets, but Jupyter-Widgets and IPyWidgets are the very same thing. To
not break compatibility with older code, the software library is still called

> `ipywidgets`.

Reviewer: These may be a bit too detailed for a basic introduction.

I feel like, for a user just getting started with Jupyter, they should learn first about the notebook concept and 1 or _maybe_ 2 apps (Notebook and Lab). Additional documentation can shed light on other projects.
